### PR TITLE
pkg/proto: adopt CodecV2 and gRPC buffer pooling

### PR DESCRIPTION
**Pull request by jzelinskie**
This replaces our existing gRPC codec (which enabled us to use vtprotobuf) with the new v2 codec interface that can support buffer pooling.

Huge thanks to engineers from Google and LinkedIn for fleshing out this work and upstreaming the necessary changes. There was a great talk from gRPConf I can share once it's uploaded to YouTube.

I did regenerate our go.mod which combined all of the indirect dependencies into one block. If there were purposely split out, I can fix it.

See the sister PR to this introducing the Codec for Vitess: https://github.com/vitessio/vitess/pull/16790

### CODE REVIEW

File: pkg/proto/dispatch/v1/01_codec.go [hunk 1]
```diff
@@ -20,38 +18,51 @@ import (
...
+			}
+			return mem.BufferSlice{mem.SliceBuffer(buf)}, nil
+		}
+		pool := mem.DefaultBufferPool()
+		buf := pool.Get(size)
+		if _, err := m.MarshalToSizedBufferVT((*buf)[:size]); err != nil {
+			pool.Put(buf)
+			return nil, err
+		}
+		return mem.BufferSlice{mem.NewBuffer(buf, pool)}, nil
```
Comment (tstirrat15) [inline pkg/proto/dispatch/v1/01_codec.go:48]: Curious - what does `NewBuffer` do here? Haven't we already reserved the relevant memory in the `pool.Get` call above?

Alternatively, is there some reading about how this approach works generally?

**Review by tstirrat15 — COMMENTED** [review]
See comment; I'll let someone more familiar with the internals give an approval.

File: pkg/proto/dispatch/v1/01_codec.go (hunk 1 above)
Comment (jzelinskie) [reply pkg/proto/dispatch/v1/01_codec.go:48]: By and large, this code is inspired by the upstream default codec: https://github.com/grpc/grpc-go/blob/v1.66.2/encoding/proto/proto.go

`mem.NewBuffer` is less about allocating bytes and more about tracking the lifetime and ownership of the buffer: https://github.com/grpc/grpc-go/blob/v1.66.2/mem/buffers.go#L83-L102

**Comment by vmg** [comment]
@jzelinskie We have quite a bit of work left before the Vitess PR is green... I would really like to see how these changes affect spicedb benchmarks. Do you have a plan to benchmark this branch? Cheers!

File: pkg/proto/dispatch/v1/01_codec.go [hunk 2]
```diff
@@ -20,38 +18,52 @@ import (
...
 	if m, ok := v.(vtprotoMessage); ok {
-		return m.MarshalVT()
-	}
-
-	if m, ok := v.(proto.Message); ok {
-		return proto.Marshal(m)
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, size)
+			n, err := m.MarshalToSizedBufferVT(buf[:size])
```
Comment (josephschorr) [inline pkg/proto/dispatch/v1/01_codec.go:37]: if we're allocating the buffer as size of `size`, why do we need to slice here and below?

Comment (jzelinskie) [reply pkg/proto/dispatch/v1/01_codec.go]: I cleaned this up a bit to try and make it more clear (and after checking guarantees from the pool implementations).
I'm being defensive for the scenario where the number of bytes written are not actually the exact size of the buffer, which looks possible based on the generated code.

**Comment by jzelinskie** [comment]
> @jzelinskie We have quite a bit of work left before the Vitess PR is green... I would really like to see how these changes affect spicedb benchmarks. Do you have a plan to benchmark this branch? Cheers!

We're getting out 2 stable releases of SpiceDB and then looking to begin testing this in some different staging environments to see how it goes.

**Review by josephschorr — APPROVED** [review]
LGTM

**Pull request merged**